### DISCUSSION TITLE
How to allow only ISO8601 timestamp in query parameter

### DISCUSSION THREAD

**Discussion by ahmedrafayat** [fastapi/fastapi#9938] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from datetime import datetime
from fastapi import FastAPI, Query


app = FastAPI()

@app.get("/test")
def test_route(time_1: datetime = Query()):
    print(time_1)  # This is always a datetime
    return time_1
```

##### Case 1:
* Open browser and request to `/test?time_1=123`
* datetime is logged in the console "1970-01-01T00:02:03+00:00"
* Returned response is "1970-01-01T00:02:03+00:00"

##### Case 2:
* Open browser and request to `/test?time_1=2023-07-20T21:33:33.232Z`
* datetime is logged in the console "2023-07-20T21:33:33.232000+00:00"
* Returned response is "2023-07-20T21:33:33.232000+00:00"

This is fine, and expected. But I need to not allow case 2, and throw error if case 1. Meaning if time_1 is not a valid ISOFormat, I want to throw error.  These are the solutions I tried.

### Solution I tried:

```
from datetime import datetime
from fastapi import Depends, FastAPI, Query
from pydantic import BaseModel, validator


app = FastAPI()



class QueryParams(BaseModel):
    time_1: str = Query()
    
    @validator('time_1')
    def validator_datetime(cls, value):
        try:
            return datetime.strptime(value, "%Y-%m-%dT%H:%M%z")  # special date format without seconds and microseconds
        except ValueError as error:
            raise ValueError("Invalid datetime") from error

@app.get("/test")
def test_route(time_1: QueryParams = Depends()):
    print(time_1)
    return time_1
```

This validates the query param, but raising ValueError here returns "Internal Server Error". So needed to add an exception handler, and so the updated code:

```
from datetime import datetime
from fastapi import Depends, FastAPI, Query
from fastapi.encoders import jsonable_encoder
from fastapi.responses import JSONResponse
from pydantic import BaseModel, validator


app = FastAPI()

@app.exception_handler(ValueError)
def handle_value_error(request, exc):
    return JSONResponse(
        status_code=422,
        content={"detail": jsonable_encoder(exc.errors()), "status": 422},
    )

class QueryParams(BaseModel):
    time_1: str = Query()
    
    @validator('time_1')
    def validator_datetime(cls, value):
        try:
            return datetime.strptime(value, "%Y-%m-%dT%H:%M%z")  # special date format without seconds and microseconds
        except ValueError as error:
            raise ValueError("Invalid datetime") from error

@app.get("/test")
def test_route(time_1: QueryParams = Depends()):
    print(time_1)
    return time_1
```

#### Problem with Solution:
1. The returned Json response for Case 1 now is
```
{
  "detail": [
    {
      "loc": [
        "time_1"
      ],
      "msg": "Invalid datetime",
      "type": "value_error"
    }
  ],
  "status": 422
}
```
but expected
```
{
  "detail": [
    {
      "loc": [
       "query",   // This is missing
        "time_1"
      ],
      "msg": "Invalid datetime",
      "type": "value_error"
    }
  ],
  "status": 422
}
```

### Operating System

Linux, macOS

### Operating System Details

Enviroments:

1. Macos 
2. Docker container running in Docker desktop in MacOS

### FastAPI Version

0.81.0

### Python Version

3.9.16

**Comment by jgould22** [comment]
What happens if you return a `pydantic.ValidationError` in `validator_datetime`?

**Reply by ahmedrafayat** [reply]
`pydantic.ValidationError` expects 3 arguments. I would have to manually build them. Can you suggest how I can do that?

**Comment by YuriiMotov** [comment, marked as answer]
You can implement this using `PlainValidator`:

```python
from datetime import datetime
from typing import Annotated, TypeAlias
from fastapi import FastAPI, Query
from pydantic import PlainValidator

app = FastAPI()

def validate_date_time(value: str):
    try:
        return datetime.strptime(value, "%Y-%m-%dT%H:%M%z")
    except ValueError as error:
        raise ValueError("Invalid datetime") from error

MyDateTime: TypeAlias = Annotated[datetime, PlainValidator(validate_date_time)]

@app.get("/test")
def test_route(time_1: Annotated[MyDateTime, Query(description="ISO8601 timestamp")]):
    print(time_1)  # This is always a datetime
    return time_1

```
